### PR TITLE
BODS-6340 ETL_tracks_data

### PR DESCRIPTION
**Pull request by SamerIssam**
Adding tracks to the ETL

### CODE REVIEW

File: transit_odp/pipelines/pipelines/dataset_etl/transform.py [hunk 1]
```diff
@@ -52,6 +54,8 @@ def __init__(self, feed_parser):
     def transform(self, extracted_data: ExtractedData) -> TransformedData:
         services = extracted_data.services.iloc[:]  # make transform immutable
         journey_patterns = extracted_data.journey_patterns.copy()
+        journey_pattern_tracks = extracted_data.journey_pattern_tracks.copy
```
Comment (NitishAroraKPMG) [inline transit_odp/pipelines/pipelines/dataset_etl/transform.py:57]: extracted_data.journey_pattern_tracks.copy()

function parenthesis are missing

File: transit_odp/pipelines/pipelines/dataset_etl/utils/dataframes.py [hunk 2]
```diff
@@ -296,6 +298,88 @@ def df_to_vehicle_journeys(df: pd.DataFrame) -> Iterator[VehicleJourney]:
...
+
+    Returns:
+        pd.DataFrame: The merged DataFrame with track and vehicle journey information.
+    """
+    # Explode the 'jp_ref' and 'rs_ref' columns along with their corresponding 'rs_order'
+    tracks_map_extended = tracks_map.explode("jp_ref").explode(["rs_ref", "rs_order"])
+
+    tracks_columns_to_keep = ["rl_ref", "rs_ref", "rl_order", "id"]
+    tracks = tracks[tracks_columns_to_keep]
+    tracks = tracks.copy()
```
Comment (NitishAroraKPMG) [inline transit_odp/pipelines/pipelines/dataset_etl/utils/dataframes.py:340]: what is the purpose of the copy() if we are assigning to same variable

File: transit_odp/pipelines/pipelines/dataset_etl/utils/transform.py [hunk 3]
```diff
@@ -940,3 +940,23 @@ def merge_flexible_jd_with_jp(
         journey_patterns, how="left", on=["file_id", "journey_pattern_id"]
     )
     return journey_details
+
+
+def transform_geometry_tracks(df):
+    if df.empty:
+        return df
+    if "geometry" in df.columns:
```
Comment (NitishAroraKPMG) [inline transit_odp/pipelines/pipelines/dataset_etl/utils/transform.py:948]: if not df.empty and ""geometry" in df.columns:
 .....

File: transit_odp/pipelines/pipelines/dataset_etl/utils/transform.py [hunk 4]
```diff
@@ -940,3 +940,23 @@ def merge_flexible_jd_with_jp(
         journey_patterns, how="left", on=["file_id", "journey_pattern_id"]
     )
     return journey_details
+
+
+def transform_geometry_tracks(df):
+    if df.empty:
```
Comment (NitishAroraKPMG) [inline transit_odp/pipelines/pipelines/dataset_etl/utils/transform.py:946]: remove it

File: transit_odp/pipelines/pipelines/dataset_etl/utils/transform.py [hunk 5]
```diff
@@ -940,3 +940,23 @@ def merge_flexible_jd_with_jp(
         journey_patterns, how="left", on=["file_id", "journey_pattern_id"]
     )
     return journey_details
+
+
+def transform_geometry_tracks(df):
```
Comment (NitishAroraKPMG) [inline transit_odp/pipelines/pipelines/dataset_etl/utils/transform.py:945]: add documentation for the function

File: transit_odp/pipelines/pipelines/dataset_etl/utils/transform.py [hunk 6]
```diff
@@ -940,3 +940,23 @@ def merge_flexible_jd_with_jp(
...
+        df["geometry"] = df["geometry"].apply(
+            lambda points: LineString(
+                [(float(long), float(lat)) for long, lat in points]
+            )
+        )
+    return df
+
+
+def add_tracks_sequence(df):
+    if df.empty:
```
Comment (NitishAroraKPMG) [inline transit_odp/pipelines/pipelines/dataset_etl/utils/transform.py:958]: add documentation

File: transit_odp/timetables/extract.py [hunk 7]
```diff
@@ -316,6 +318,103 @@ def extract_journey_patterns(self):
 
         return journey_patterns, jp_to_jps
 
+    def extract_journey_rout_link(self):
```
Comment (NitishAroraKPMG) [inline transit_odp/timetables/extract.py:321]: rename it to extract_journey_route_link

File: transit_odp/timetables/extract.py [hunk 8]
```diff
@@ -316,6 +318,103 @@ def extract_journey_patterns(self):
...
+        # Get services and journey patterns
```
Comment (NitishAroraKPMG) [inline transit_odp/timetables/extract.py:322]: Add documentation

File: transit_odp/timetables/extract.py [hunk 9]
```diff
@@ -316,6 +318,103 @@ def extract_journey_patterns(self):
...
+            jp_route_ref.groupby("route_ref")["pattern_id"].apply(list).reset_index()
+        )
+        df_vj_grouped = df_vj_grouped.rename(columns={"pattern_id": "jp_ref"})
+
+        # Get routes and create a route reference link dictionary
+        routes = self.doc.get_route()
+        route_ref_link = {}
+
+        for route in routes:
+            route.text
```
Comment (NitishAroraKPMG) [inline transit_odp/timetables/extract.py:340]: what are we trying to do with this statement

File: transit_odp/timetables/extract.py [hunk 10]
```diff
@@ -316,6 +318,103 @@ def extract_journey_patterns(self):
...
+            route_section_refs = route.get_elements_or_none(["RouteSectionRef"])
+            route_ref_ids = []
+
+            # Extract route section references and update route_ref_link dictionary
+            for route_section_ref in route_section_refs:
+                route_id = route.get("id")
```
Comment (NitishAroraKPMG) [inline transit_odp/timetables/extract.py:346]: better to get routeid in the upper loop

File: transit_odp/timetables/extract.py [hunk 11]
```diff
@@ -316,6 +318,103 @@ def extract_journey_patterns(self):
...
+                if route_section_ref and route_id:
+                    route_section_ref_text = (
+                        route_section_ref.text if route_section_ref else ""
+                    )
+                    route_ref_ids.append(route_section_ref_text)
+
+            route_ref_link.update({route.get("id"): route_ref_ids})
```
Comment (NitishAroraKPMG) [inline transit_odp/timetables/extract.py:353]: we are already having route_id

File: transit_odp/timetables/extract.py [hunk 12]
```diff
@@ -316,6 +318,103 @@ def extract_journey_patterns(self):
...
+
+        # Convert route_ref_link dictionary to DataFrame
+        route_map = pd.DataFrame(
```
Comment (NitishAroraKPMG) [inline transit_odp/timetables/extract.py:356]: let us move this section to new function for better understanding of the functin

File: transit_odp/timetables/loaders.py [hunk 13]
```diff
@@ -213,9 +228,39 @@ def load_vehicle_journeys(self, service_patterns):
             vehicle_journeys["id"] = pd.Series(
                 (obj.id for obj in created), index=vehicle_journeys.index
             )
-
         return vehicle_journeys
 
+    def load_journey_tracks(self):
+        create_or_update = []
```
Comment (NitishAroraKPMG) [inline transit_odp/timetables/loaders.py:234]: add documenration

File: transit_odp/timetables/loaders.py [hunk 14]
```diff
@@ -213,9 +228,39 @@ def load_vehicle_journeys(self, service_patterns):
...
+                defaults=track_dict,
+            )
+            create_or_update.append(obj)
+
+        tracks["id"] = pd.Series(
+            (obj.id for obj in create_or_update), index=tracks.index
+        )
+        return tracks
+
+    def load_vj_tracks(self, tracks, vehicle_journeys, tracks_map):
```
Comment (NitishAroraKPMG) [inline transit_odp/timetables/loaders.py:251]: add documenration

File: transit_odp/timetables/transformers.py [hunk 15]
```diff
@@ -83,6 +88,17 @@ def transform(self) -> TransformedData:
...
             ~stop_points.index.isin(flexible_stop_points.index)
         ]
+
+        if not journey_pattern_tracks.empty:
+            journey_pattern_tracks = transform_geometry_tracks(journey_pattern_tracks)
+            journey_pattern_tracks = add_tracks_sequence(journey_pattern_tracks)
+
+        if not route_map.empty:
+            # Add 'rs_order' column to tracks_map indicating the order of each stop reference
+            route_map["rs_order"] = route_map["rs_ref"].apply(
```
Comment (NitishAroraKPMG) [inline transit_odp/timetables/transformers.py:98]: in this also, do we need to copy as we are not renaming, just changing column

**Review by NitishAroraKPMG — DISMISSED** [review]

**Review by ankitmahajan1 — APPROVED** [review]
approved

**Pull request merged**